add error message for missing strata

chrisorwa opened the pull request.

blackorwa committed:
add error message for missing strata

blackorwa committed:
run document() for est.incidence.by

blackorwa committed:
add any() to is.element()

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -101,6 +101,27 @@ est.incidence.by <- function(
+    cli::cli_abort(
+      class = "missing_var",
+      message = c(
+        "Can't find the column {.var {strata}} in {.envvar pop_data}.",
```
please revise this inline markup; `pop_data` is a [function argument](https://cli.r-lib.org/reference/inline-markup.html#:~:text=arg%20for%20a%20function%20argument), not an environment variable (see https://cli.r-lib.org/reference/inline-markup.html)

chrisorwa replied:
This is well-noted.

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -101,6 +101,27 @@ est.incidence.by <- function(
+        "Can't find the column {.var {strata}} in {.envvar pop_data}.",
+        if (length(strata_var) > 0) {
+          "i" <- "Did you mean {.var {strata_var}}."
+        }
```
please add a test that covers this case

chrisorwa replied:
This has been implemented.

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -101,6 +101,27 @@ est.incidence.by <- function(
     return(to_return)
   }
 
+  if (!any(is.element(strata,pop_data %>% names()))) {
```
please always format according to the style guide (add a space after the comma in `strata,pop_data`, etc)

chrisorwa replied:
The change has been made.

d-morrison reviewed (changes requested):
Thanks for this PR; please address comments and add tests for the code lines in this PR.

blackorwa committed:
add test

blackorwa committed:
Merge branch 'main' into add-user-error-message-for-missing-strata

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -101,6 +101,27 @@ est.incidence.by <- function(
     return(to_return)
   }
 
+  if (!any(is.element(strata, pop_data %>% names()))) {
```
should `any()` be `all()`?

chrisorwa replied:
This should be `all()`. The change has been made.

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -101,6 +101,27 @@ est.incidence.by <- function(
+      grep(
+        x = pop_data %>% names(),
+        value = TRUE,
+        pattern = strata,
```
what happens here if `strata` has multiple entries in it? for example, `strata = c("Country", "catch")`? Please check this scenario and adjust this code if needed (probably, try to match each missing stratum variable separately).

chrisorwa replied:
This is in order and adds complexity to Issue #233. For example, if `strata = c("age", "Country")`, I had opted to use regex `paste0("^", strata)` to capture only columns that begin with `age` to avoid and avoid words such as **village**.

**Addendum:**
Never mind, I found a solution to it.

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -101,6 +101,27 @@ est.incidence.by <- function(
+      message = c(
+        "Can't find the column {.arg {strata}} in {.arg pop_data}.",
+        if (length(strata_var) > 0) {
+          "i" <- "Did you mean {.var {strata_var}}."
```
Please end questions with a question mark instead of a period.

chrisorwa replied:
This has been implemented.

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -101,6 +101,27 @@ est.incidence.by <- function(
+    cli::cli_abort(
+      class = "missing_var",
+      message = c(
+        "Can't find the column {.arg {strata}} in {.arg pop_data}.",
```
Please inform the user that this error relates to their input for the `strata` argument, so they know which input to fix.

chrisorwa replied:
This has been implemented.

d-morrison reviewed (changes requested):
Getting closer; please see inline comments.

blackorwa committed:
Merge branch 'main' into add-user-error-message-for-missing-strata

blackorwa committed:
modify error messaging

blackorwa committed:
Merge branch 'add-user-error-message-for-missing-strata' of https://github.com/UCD-SERG/serocalculator into add-user-error-message-for-missing-strata

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -101,6 +101,36 @@ est.incidence.by <- function(
 
+  if (!all(is.element(strata, pop_data %>% names()))) {
+    # accommodate multiple strata
+    combined_pattern <- paste(strata, collapse = "|")
```
This approach seems sub-optimal, for two reasons:

1. we won't be able to tell which columns partially match each specific missing element of `strata`.
2. we didn't remove the elements of `strata` that already exactly match columns in `pop_data`

More generally, complicated regular expressions are a recipe for bugs; if you think you need to use `|` in a regex, please carefully consider alternative approaches. As always, we want to keep the code as simple as possible; for example, in this case, instead of pasting `strata` together, we could instead individually grep each element of `strata` that isn't exactly in `pop_data`, using a `for` loop, or `lapply()`, or `purrr::map()`, etc. 

Then, we can report the possible partial matches for each missing element of `strata`; the markup capabilities of `cli_abort()` enable us to provide this information in an orderly, structured format. Please give all this a try, but if you get stuck, just let me know and I'll provide more detailed instructions.

chrisorwa replied:
I have made changes to this approach to capture full match, partial match and no match.

d-morrison commented on tests/testthat/test-est.incidence.R:
```diff
@@ -79,3 +79,36 @@ test_that("`est.incidence()` produces expected results", {
+  curve <- load_curve_params("https://osf.io/download/rtw5k/")
+
+  expect_error(
+    object = est.incidence.by(
```
why are we testing `est.incidence.by()` in the test file for `est.incidence()`? Please be careful to keep these tests properly organized.

d-morrison replied:
please add a test in which `strata` contains multiple elements that don't exactly match the columns of `pop_data`

chrisorwa replied:
I have separated `est.incidence` and `est.incidence.by` into different scripts. In addition, I have added a test for multiple `strata` in `est.incidence.by`.

d-morrison reviewed (changes requested):
Thanks for the hard work on this PR; please see inline comments.

blackorwa committed:
make changes to tests and est.incidence.by

blackorwa committed:
add purrr to DESCRIPTION

blackorwa committed:
modify error message

blackorwa committed:
correct errors

blackorwa committed:
remove testthat from DESCRIPTION

blackorwa committed:
Merge branch 'main' into add-user-error-message-for-missing-strata

d-morrison committed:
in progress

d-morrison reviewed (changes requested):
Looks like this PR is still in-progress?

- [x] Please address the lint errors and restore the files that seem to have been accidentally deleted. 

- [x] Before re-requesting review, please look through [Files changed](https://github.com/UCD-SERG/serocalculator/pull/227/files) and add at least one comment per changed file: briefly explain what's been changed and why.

blackorwa committed:
Merge branch 'main' into add-user-error-message-for-missing-strata

blackorwa committed:
lint file

blackorwa committed:
lint tests

blackorwa committed:
Merge branch 'main' into add-user-error-message-for-missing-strata

chrisorwa commented on R/est.incidence.by.R:
```diff
@@ -1,14 +1,22 @@
-#' @param curve_strata_varnames A subset of `strata`. Values must be variable names in `curve_params`. Default = "".
-#' @param noise_strata_varnames A subset of `strata`. Values must be variable names in `noise_params`. Default = "".
-#' @param num_cores Number of processor cores to use for calculations when computing by strata. If set to more than 1 and package \pkg{parallel} is available, then the computations are executed in parallel. Default = 1L.
+#' @param pop_data a [data.frame] with cross-sectional serology data per
+#' antibody and age, and additional columns corresponding to
+#' each element of the `strata` input
+#' @param strata a [character] vector of stratum-defining variables.
+#' Values must be variable names in `pop_data`.
+#' @param curve_strata_varnames A subset of `strata`.
+#' Values must be variable names in `curve_params`. Default = "".
+#' @param noise_strata_varnames A subset of `strata`.
+#' Values must be variable names in `noise_params`. Default = "".
+#' @param num_cores Number of processor cores to use for
+#' calculations when computing by strata. If set to
+#' more than 1 and package \pkg{parallel} is available,
+#' then the computations are executed in parallel. Default = 1L.
```
Reduce sentence length to 80 characters to fit the linting requirement.

chrisorwa replied:
Reduce sentences to 80 characters (linting requirement).

chrisorwa commented on R/est.incidence.by.R:
```diff
@@ -17,7 +25,8 @@
 #' If for some reason you want to use [est.incidence.by()]
 #' with no strata instead of calling [est.incidence()],
-#' you may use `NA`, `NULL`, or `""` as the `strata` argument to avoid that warning.
+#' you may use `NA`, `NULL`, or `""` as the `strata`
+#' argument to avoid that warning.
```
Reduce sentences to 80 characters to match the linting requirement.

chrisorwa commented on R/est.incidence.by.R:
```diff
@@ -26,7 +35,9 @@
-#' An object of class `"seroincidence.by"`; i.e., a list of `"seroincidence"` objects from [est.incidence()], one for each stratum, with some meta-data attributes.
+#' An object of class `"seroincidence.by"`; i.e., a list of `
+#' "seroincidence"` objects from [est.incidence()], one for each stratum,
+#' with some meta-data attributes.
```
Reduce sentences to 80 characters to match the linting requirement.

chrisorwa commented on R/est.incidence.by.R:
```diff
@@ -39,7 +50,7 @@
 #' curve <- load_curve_params("https://osf.io/download/rtw5k/") %>%
 #'   filter(antigen_iso %in% c("HlyE_IgA", "HlyE_IgG")) %>%
-#'   slice(1:100, .by = antigen_iso) # Reduce dataset for the purposes of this example
+#'   slice(1:100, .by = antigen_iso)
```
Remove comment to make the line less than 80 characters.

d-morrison replied:
instead of removing the comment, why not move it to a new line?

chrisorwa commented on DESCRIPTION:
```diff
@@ -31,7 +31,8 @@ Imports:
     tidyselect,
-    cli
+    cli,
+    purrr
```
Add `purrr` library.

chrisorwa commented on R/est.incidence.by.R:
```diff
@@ -1,14 +1,22 @@
 #' @description
-#' Function to estimate seroincidences based on cross-section serology data and longitudinal
+#' Function to estimate seroincidences based on cross-section
+#' serology data and longitudinal
```
Reduce sentences to 80 characters (linting requirement).

chrisorwa commented on R/est.incidence.by.R:
```diff
@@ -75,16 +86,19 @@ est.incidence.by <- function(
+      simplify your code and avoid this warning.",
+      "\n\n Since the `strata` argument is empty, `est.incidence.by()`
+      will return a `seroincidence` object, instead of a
+      `seroincidence.by` object.\n"
```
Reduce sentences to 80 characters (linting requirement).

d-morrison replied:
please add tests everywhere indicated by Codecov comments

chrisorwa commented on R/est.incidence.by.R:
```diff
@@ -75,16 +86,19 @@ est.incidence.by <- function(
-      is.null(strata) ||
-      setequal(strata, NA) ||
-      setequal(strata, "")
+    is.null(strata) ||
+    setequal(strata, NA) ||
+    setequal(strata, "")
```
Reduce space to meet linting requirements.

chrisorwa commented on R/est.incidence.by.R:
```diff
@@ -130,8 +144,10 @@ est.incidence.by <- function(
     warning(
-      "The `parallel` package is not installed, so `num_cores > 1` has no effect.",
-      "To install `parallel`, run `install.packages('parallel')` in the console."
+      "The `parallel` package is not installed,
+      so `num_cores > 1` has no effect.",
+      "To install `parallel`, run `install.packages('parallel')`
+      in the console."
```
Reduce sentences to 80 characters (linting requirement).

chrisorwa commented on R/est.incidence.by.R:
```diff
@@ -142,7 +158,8 @@ est.incidence.by <- function(
     if (verbose) {
-      message("Setting up parallel processing with `num_cores` = ", num_cores, ".")
+      message("Setting up parallel processing with
+              `num_cores` = ", num_cores, ".")
```
Reduce sentences to 80 characters (linting requirement).

chrisorwa commented on R/est.incidence.by.R:
```diff
@@ -158,10 +175,10 @@ est.incidence.by <- function(
       .libPaths(libPaths)
-      require(serocalculator) # note - this gets out of sync when using load_all() in development
+      require(serocalculator) # note - this gets out of sync" %>%
+      "when using load_all() in development"
```
Reduce sentences to 80 characters (linting requirement).

d-morrison replied:
it looks like the second half of that comment got turned into code; why?

chrisorwa commented on tests/testthat/_snaps/est.incidence.by.md:
```diff
@@ -0,0 +1,11 @@
+            <dbl>          <dbl>  <dbl>  <dbl>  <dbl>    <dbl>   <dbl>      <int>
+      1       0.1          0.133 0.0216 0.0962  0.182     0.95   -261.          4
+      # i 2 more variables: antigen.isos <chr>, nlm.convergence.code <ord>
+
```
A redo of the typhoid data test using test snapshot.

blackorwa committed:
increment version

chrisorwa commented on tests/testthat/test-as_curve_params.R:
This file has been moved to PR #228 . This file was originally part of the `as_noise_params` PR and got to this PR by mistake. I have moved it back to the noise PR.

d-morrison committed:
line break

d-morrison commented on DESCRIPTION:
```diff
@@ -1,7 +1,7 @@
 Type: Package
 Title: Estimating Infection Rates from Serological Data
-Version: 1.2.0.9001
+Version: 1.3.0
```
please increment dev version instead of release version

d-morrison commented on NEWS.md:
```diff
@@ -1,4 +1,7 @@
-# serocalculator (development version)
+# serocalculator 1.3.0
```
use dev version

d-morrison replied:
see comment in previous review; this change still needs to be reverted

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -26,7 +35,9 @@
 #' @return
 #' * if `strata` has meaningful inputs:
-#' An object of class `"seroincidence.by"`; i.e., a list of `"seroincidence"` objects from [est.incidence()], one for each stratum, with some meta-data attributes.
+#' An object of class `"seroincidence.by"`; i.e., a list of `
```
please move linebreak to before the backtick

d-morrison committed:
Merge branch 'main' into add-user-error-message-for-missing-strata

d-morrison commented on NEWS.md:
```diff
@@ -1,4 +1,7 @@
-# serocalculator (development version)
+# serocalculator 1.3.0
+* Add test for `est.incidence` 
```
est.incidence.*by*?

d-morrison replied:
and what behavior is being tested?

d-morrison commented on NEWS.md:
```diff
@@ -1,4 +1,7 @@
+# serocalculator 1.3.0
+* Add test for `est.incidence` 
+
+* Add option to test `strata` in `pop_data`
```
is this a user-facing option? not sure what this means.

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -1,14 +1,22 @@
-#'
 #' @description
-#' Function to estimate seroincidences based on cross-section serology data and longitudinal
+#' Function to estimate seroincidences based on cross-section
```
cross-section*al*

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -75,16 +86,19 @@ est.incidence.by <- function(
     warning(
```
please update from `warning()` to `cli_warn()`

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -142,7 +158,8 @@ est.incidence.by <- function(
 
     if (verbose) {
-      message("Setting up parallel processing with `num_cores` = ", num_cores, ".")
+      message("Setting up parallel processing with
```
Please update from `message()` to `cli_inform()`

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -158,10 +175,10 @@ est.incidence.by <- function(
     parallel::clusterEvalQ(cl, {
       .libPaths(libPaths)
-      require(serocalculator) # note - this gets out of sync when using load_all() in development
+      require(serocalculator) # note - this gets out of sync" %>%
```
why add `%>%` at the end of this line? what does it do?

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -158,10 +175,10 @@ est.incidence.by <- function(
+      "when using load_all() in development"
       require(dplyr)
     })
-
```
why?

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -225,7 +235,8 @@ est.incidence.by <- function(
             )
           )
       }
-    } %>% system.time() -> time
```
we don't want to lose the `system.time()` call; we need it below for verbose messaging. 

I understand that this was changed to avoid the right-hand assignment operator `->`, but there needs to be a left-hand assignment at the beginning of this whole expression instead.

d-morrison commented on tests/testthat/_snaps/as_curve_params.md:
why deleted? if accidental, please revert to `main` branch version.

d-morrison commented on tests/testthat/test-est.incidence.by.R:
```diff
@@ -0,0 +1,81 @@
+test_that(
+  "est.incidence() produces expected results for typhoid data",
```
est.incidence.*by*()?

d-morrison commented on tests/testthat/test-est.incidence.by.R:
```diff
@@ -0,0 +1,81 @@
+  }
+)
+
+test_that("`est.incidence()` produces expected results", {
```
ditto

d-morrison commented on tests/testthat/test-est.incidence.by.R:
```diff
@@ -0,0 +1,81 @@
+    # set start
+    start <- .05
+
+    typhoid_results <- est.incidence(
```
the function being tested (`est.incidence()`) doesn't match the file name (`test-est-incidence.by`). Please sort this out.

d-morrison commented on tests/testthat/test-est.incidence.by.R:
```diff
@@ -0,0 +1,81 @@
+    slice_head(n = 100)
+
+
+  est_true <- est.incidence(
```
ditto

d-morrison commented on tests/testthat/test-est.incidence.by.R:
```diff
@@ -0,0 +1,81 @@
+    slice_head(n = 100)
+
+
+  est_false <- est.incidence(
```
ditto

d-morrison commented on tests/testthat/test-as_curve_params.R:
see above

d-morrison commented on tests/testthat/test-est.incidence.by.R:
```diff
@@ -0,0 +1,81 @@
+        start = start
+      )
+
+    expect_snapshot(x = typhoid_results)
```
please add an additional test using `expect_snapshot_value()` or `expect_snapshot_file()` (please think through which one seems best to use).

d-morrison committed:
increment version

d-morrison reviewed (changes requested):
Thanks for working on this! Please see comments. Not sure why codecov didn't throw an error; please increase the patch coverage.

blackorwa committed:
Merge branch 'main' into add-user-error-message-for-missing-strata

blackorwa committed:
Merge branch 'main' into add-user-error-message-for-missing-strata

blackorwa committed:
correct conflicts

blackorwa committed:
requested changes

blackorwa committed:
push additional changes to tests

blackorwa committed:
Merge branch 'main' into add-user-error-message-for-missing-strata

blackorwa committed:
Merge branch 'main' into add-user-error-message-for-missing-strata

Commit:
reduce tests

Commit:
Merge branch 'add-user-error-message-for-missing-strata' of https://github.com/UCD-SERG/serocalculator into add-user-error-message-for-missing-strata

blackorwa committed:
reduce num_cores to 1

blackorwa committed:
add snapshots

blackorwa committed:
linting

blackorwa committed:
correct tests

blackorwa committed:
Increment version number to 1.2.0.9019

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -26,7 +35,9 @@
 #' * if `strata` has meaningful inputs:
-#' An object of class `"seroincidence.by"`; i.e., a list of `"seroincidence"` objects from [est.incidence()], one for each stratum, with some meta-data attributes.
+#' An object of class `"seroincidence.by"`; i.e., a list of `
+#' "seroincidence"` objects from [est.incidence()], one for each stratum,
```
```suggestion
#' An object of class `"seroincidence.by"`; i.e., a list of
#' `"seroincidence"` objects from [est.incidence()], one for each stratum,
```

d-morrison commented on R/est.incidence.by.R:
```diff
@@ -39,7 +50,8 @@
 #' curve <- load_curve_params("https://osf.io/download/rtw5k/") %>%
 #'   filter(antigen_iso %in% c("HlyE_IgA", "HlyE_IgG")) %>%
-#'   slice(1:100, .by = antigen_iso) # Reduce dataset for the purposes of this example
+#'  # Reduce dataset for the purposes of this example
```
```suggestion
#'  # Reduce dataset for the purposes of this example:
```

d-morrison committed:
more edits

d-morrison committed:
in progress

Merge branch 'add-user-error-message-for-missing-strata' into simplifications

# Conflicts:
#	R/est.incidence.by.R
#	tests/testthat/test-est.incidence.by.R
#	vignettes/.gitignore

d-morrison committed:
lots

d-morrison committed:
more cleanup

d-morrison committed:
update documentation

d-morrison committed:
- added {and} package

d-morrison committed:
cleanup

d-morrison committed:
add `glue` and `stringr`

d-morrison committed:
add test to `check_strata()`

d-morrison committed:
remove redundant example data

d-morrison committed:
wrap

d-morrison committed:
Merge branch 'simplifications' into use-test-data

d-morrison committed:
Merge pull request #316 from UCD-SERG/use-test-data

Use test data in simplifications

d-morrison committed:
remove redundant files

d-morrison committed:
delete unused files

d-morrison committed:
Merge remote-tracking branch 'origin/use-test-data' into simplifications

d-morrison committed:
fixes

d-morrison committed:
Merge branch 'lint-changed-faster' into simplifications

d-morrison committed:
Merge pull request #318 from UCD-SERG/use-test-data

clean up tests

d-morrison committed:
Merge branch 'main' into add-user-error-message-for-missing-strata

d-morrison committed:
Merge branch 'add-user-error-message-for-missing-strata' into simplifications

d-morrison committed:
fix

d-morrison committed:
increase snapshot tolerance

d-morrison committed:
lint

d-morrison committed:
document

d-morrison committed:
Merge remote-tracking branch 'origin/main' into simplifications

d-morrison committed:
Increment version number to 1.2.0.9022

d-morrison committed:
remove duplicate line

d-morrison committed:
try to fix lint

d-morrison committed:
Merge branch 'main' into add-user-error-message-for-missing-strata

d-morrison committed:
more linting

d-morrison committed:
Merge branch 'add-user-error-message-for-missing-strata' into simplifications

d-morrison committed:
lint

d-morrison committed:
Merge branch 'simplifications' of https://github.com/UCD-SERG/serocalculator into simplifications

d-morrison committed:
new test to hopefully complete patch coverage

d-morrison committed:
skip parallel cores test if only one core available

d-morrison committed:
added test of verbose output

d-morrison committed:
cleanup

d-morrison committed:
test verbose and multicore

d-morrison committed:
more checks

d-morrison committed:
cleaning up time output

d-morrison committed:
fixes

d-morrison committed:
formatting

d-morrison committed:
Merge branch 'main' into add-user-error-message-for-missing-strata

d-morrison committed:
Merge branch 'add-user-error-message-for-missing-strata' into simplifications

d-morrison reviewed (changes requested):
Thanks for this work; please see comments below and suggested edits in https://github.com/UCD-SERG/serocalculator/pull/245

blackorwa committed:
Update R/est.incidence.by.R

Co-authored-by: Douglas Ezra Morrison <demorrison@ucdavis.edu>

blackorwa committed:
Update R/est.incidence.by.R

Co-authored-by: Douglas Ezra Morrison <demorrison@ucdavis.edu>

blackorwa committed:
additional changes

blackorwa committed:
Merge branch 'add-user-error-message-for-missing-strata' of https://github.com/UCD-SERG/serocalculator into add-user-error-message-for-missing-strata

blackorwa committed:
change version number

blackorwa committed:
make changes to NEWS.md

blackorwa committed:
Increment version number to 1.2.0.9022

d-morrison committed:
Merge branch 'add-user-error-message-for-missing-strata' into simplifications

blackorwa committed:
Merge pull request #245 from UCD-SERG/simplifications

more edits to est.incidence.by()

d-morrison committed:
Merge branch 'main' into add-user-error-message-for-missing-strata

blackorwa committed:
make changes

blackorwa committed:
Merge branch 'add-user-error-message-for-missing-strata' of https://github.com/UCD-SERG/serocalculator into add-user-error-message-for-missing-strata

d-morrison reviewed (approved):
Looks good, thanks!
